pkg/sql/conn_executor.go

```diff
@@ -3113,6 +3112,79 @@ func (ex *connExecutor) handleWaitingForDescriptorIDGeneratorMigration(ctx conte
 	return ex.resetTransactionOnSchemaChangeRetry(ctx)
 }
 
+func (ex *connExecutor) waitOneVersionForNewVersionDescriptorsWithoutJobs(
+	descIDsInJobs catalog.DescriptorIDSet,
+) error {
+	withNewVersion, err := ex.extraTxnState.descCollection.GetOriginalPreviousIDVersionsForUncommitted()
+	if err != nil {
+		return err
+	}
+	for _, idVersion := range withNewVersion {
+		if descIDsInJobs.Contains(idVersion.ID) {
+			continue
+		}
+		if _, err := WaitToUpdateLeases(ex.Ctx(), ex.planner.LeaseMgr(), idVersion.ID); err != nil {
+			// In most cases (normal schema changes), deleted descriptor should have
+			// been handled by jobs. So, normally we won't hit into the situation of
+			// wait for one version of a deleted descriptor. However, we need catch
+			// ErrDescriptorNotFound here because we have a special case of descriptor
+			// repairing where we delete descriptors directly and never record the ids
+			// in jobs payload or details.
+			if errors.Is(err, catalog.ErrDescriptorNotFound) {
+				continue
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func (ex *connExecutor) descIDsInJobs() (catalog.DescriptorIDSet, error) {
+	// Get descriptor IDs from legacy schema changer jobs.
+	var descIDsInJobs catalog.DescriptorIDSet
+	if err := ex.extraTxnState.jobsInfo.forEachJobToCreate(func(jobRecord *jobs.Record) error {
+		for _, descID := range jobRecord.DescriptorIDs {
```

Comment:
Does this cause problems for other types of jobs than schema changes like gc jobs or changefeeds? can we make sure the job has a relevant type where we know there will be waiting? Maybe rename this function to reference schema change jobs, and describe on this function why those schema change jobs are special?

Reply:
I think it'd be good to describe in commentary here the semantics of this field for the different types of schema changes.

Reply:
jobs like gc and changefeeds are fine. I renamed it to `descIDsInSchemaChangeJobs` as suggested and add more comments. I also move the iteration on `DescriptorIDs` into the `switch` block so that we only use these ids when it's actually a schema change job.